doc/Neighbor/Neighbor_Refresh_HLD.md

```diff
@@ -0,0 +1,321 @@
+```
+
+Schema:
+
+```
+; New table
+; Has the ARP/ND entries that needs to be refreshed periodically
+
+key           = NEIGH_REFRESH_TABLE:ifname:IPprefix
+
+; field       = value
+neigh         = 12HEXDIG           ; mac address of the neighbor
+family        = "IPv4" / "IPv6"    ; address family
+state         = 1*10DIGIT          ; Neighbor state in Kernel
+flags         = 1*10DIGIT          ; Neighbor flag info in Kernel
+
+Example:
+127.0.0.1:63792> hgetall "NEIGH_REFRESH_TABLE:Ethernet25:20.20.20.1"
+ 1) "neigh"
+ 2) "80:a2:35:26:0c:61"
+ 3) "family"
+ 4) "IPv4"
+ 5) "state"
+ 6) "2"
+ 7) "flags"
+ 8) "0"
+127.0.0.1:63792>
+```
+
+### 3.2.3 STATE DB
+Not applicable
+
+### 3.2.4 ASIC DB
+Not applicable
+
+### 3.2.5 COUNTER DB
+Not applicable
+
+### 3.2.6 Configuration save and restart
+Not applicable
+
+## 3.3 Switch State Service Design
+
+### 3.3.1 Orchestration Agent
+
+No Change in Orchangent / NeighOrch.
+
+### 3.3.2 NeighSyncd
+
+Neighsyncd is responsible for syncing the kernel neighbor table to the hardware via the APP_DB and OrchAgents. Neighsyncd listens on netlink events (RTM_NEWNEIGH, RTM_DELNEIGH) and creates/deletes NEIGH_TABLE entries in APP_DB.
+
+Existing functionality of neighsyncd is retained as it is. In addition to managing NEIGH_TABLE entries in APP_DB, neighsyncd will also add the details of the neighbor into NEIGH_REFRESH_TABLE.
+
+
+### 3.3.2 Nbrmgrd
+
+Nbrmgrd in addition to its existing responsibilities will now have to take care of the following Neighbor Refresh resonsibilities.
+
+- listens to NEIGH_REFRESH_TABLE and manage a local cache of neighbor entries
+- subscribe to redis-db interface CONFIG_DB tables to gather the data required to send the ARP refresh packets
+- subscribe to STATE_DB FDB table to manage the mac address table local cache
+- periodically refresh ARP/ND by sending ARP request pkt / NS pkt
+
+
+#### 3.3.2.1 Local Cache
+
+Neighbor Refresh Cache:
+
+- Add neighbor entries to cache when the entry is added to NEIGH_REFRESH_TABLE
+    - All Dynamically learned neighbor entries [ARP, ND (Global, LinkLocal)]
+    - All Static neighbor entries (MAC can be dynamic)
+- Following Entries will not be added to the neighbor cache
+    - Neighbors learned from “eth0” interface
+    - Neighbors learned from BGP/EVPN MAC/IP type-2 route 
+- Remove entries from cache when the entry is deleted from NEIGH_REFRESH_TABLE
+
+```
+v4/v6 Neighbors Cache [map]
+    
+    Key     : IP Address + InterfaceName [Phy/PortChannel/Vlan/Sag]
+
+    Value
+            : MAC Address
+            : IP address family
+            : State  (Reachable/Stale/Failed)
+            : Kernel Flags (For NTF_EXT_LEARNED)
+            : Timestamp (Entry creation/last refresh)
+```
+
+Interface Cache:
+
+- Interface IP address will be used in building the ARP (Sender IP address) and NS (Soure IP) packets
+- Subscribe to redis-db tables
+    - IP address 
+      - CONFIG_DB: INTERFACE, VLAN_INTERFACE, SAG_INTERFACE, PORTCHANNEL_INTERFACE
```

Comment:
For the regular L3 interfaces (non-VLAN interfaces), why cant we patch the kernel to refresh the nbr when the state is transitioned from reachable to stale? do we need to add those nbrs as well into the refresh Q. Kernel level refresh may be faster and efficient. any thoughts?

Reply:
One way is to change the state transition from reachable to stale to reachable to delay/probe, so that the kernel state machine takes care of pkt tx, retry etc. However this introduces state transition which further have to be processed by neighsyncd.

With scale, if kernel does the refresh there could be possible large number of tx and corresponding copp rx limit mismatch.

Building a local neighbor cache in one of the swss modules will help add additional functionalities other than neighbor refresh in future.